std_execution.bs

```diff
@@ -3022,3 +3039,152 @@ are all well-formed.
+
+    3. <b><code>value_t <i>sender-awaitable</i>::await_resume()</code></b>
+
+        - <i>Effects:</i> equivalent to:
+
+            <pre highlight="c++">
+            if (result_.index()) == 2)
+              rethrow_exception(std::get&lt;2>(result_));
+            if constexpr (!is_void_v&lt;value_t>)
+              return (value_t&&) std::get&lt;1>(result_);
+            </pre>
+
+### `execution::with_awaitable_senders` <b>[execution.coro_utils.with_awaitable_senders]</b> ### {#spec-execution.coro_utils.with_awaitable_senders}
+
+  1. `with_awaitable_senders`, when used as the base class of a coroutine promise type, makes senders awaitable in that coroutine type.
+
+    In addition, it provides a default implementation of `unhandled_done()` such that if a sender completes by calling `execution::set_done`, it is treated as if an uncatchable "done" exception were thrown from the <i>await-expression</i>. In practice, the coroutine is never resumed, and the `unhandled_done` of the coroutine caller's promise type is called.
+
+    <pre highlight="c++">
+    template &lt;class Promise>
+      requires is_class_v&lt;Promise> && same_as&lt;Promise, remove_cvref_t&lt;Promise>>
+      struct with_awaitable_senders {
+        template&lt;OtherPromise>
+          requires (!same_as&lt;OtherPromise, void>)
+        void set_continuation(coroutine_handle&lt;OtherPromise> h) noexcept;
+
+        coroutine_handle&lt;> continuation() const noexcept { return continuation_; }
+
+        coroutine_handle&lt;> unhandled_done() noexcept {
+          return done_handler_(continuation_.address());
+        }
+
+        template&lt;class Value>
+        <i>see-below</i> await_transform(Value&& value);
+
+       private:
+        coroutine_handle&lt;> continuation_{}; // exposition only
+        coroutine_handle&lt;> (*done_handler_)(void*) noexcept = // exposition only
+          [](void*) noexcept -> coroutine_handler&lt;> { terminate(); };
+      };
+    </pre>
+
+  2. <b>`void set_continuation(coroutine_handle<OtherPromise> h) noexcept`</b>
+
+      - <i>Effects:</i> equivalent to:
+
+        <pre highlight="c++">
+        continuation_ = h;
+        if constexpr ( requires(OtherPromise& other) { other.unhandled_done(); } ) {
+          done_handler_ = [](void* p) noexcept -> coroutine_handle&lt;> {
+            return coroutine_handle&lt;OtherPromise>::from_address(p)
+              .promise().unhandled_done();
+          };
+        }
+        </pre>
+
+  3. <b>`decltype(auto) await_transform(Value&& value)`</b>
+
+      - <i>Effects:</i> equivalent to:
+
+        <pre highlight="c++">
+        if constexpr (<i>is-awaitable</i>&lt;Value>)
+          return (Value&&) value;
+        else if constexpr (<i>single-typed-sender</i>&lt;Value>)
+          return as_awaitable(static_cast&lt;Value&&>(value), static_cast&lt;Promise&>(*this));
+        else
+          return (Value&&) value;
```

Comment:
This case would never expect to be reached in a well-formed program, right?
The `is-awaitable` concept check should in theory catch and cover all cases where the type should be awaitable.
If we get to the point here where we are returning value unchanged but it is not awaitable then the program is going to be ill-formed anyway...

Having said that, I wonder if we should be doing a stricter check than `is-awaitable` above - we do know the promise_type at this point so we can correctly construct the appropriate `coroutine_handle<P>` argument to `await_suspend()`.

Reply:
> The `is-awaitable` concept check should in theory catch and cover all cases where the type should be awaitable.

True. The question is one of diagnostics. By returning the non-awaitable thing unchanged, the user will get a compiler error telling them that they are trying to await a thingy that isn't awaitable. That's a much better usability story.

> we do know the promise_type at this point so we can correctly construct the appropriate `coroutine_handle<P>` argument to `await_suspend()`.

That's a fair point.

Reply:
> I wonder if we should be doing a stricter check than `is-awaitable above` - we do know the promise_type at this point so we can correctly construct the appropriate `coroutine_handle<P>` argument to `await_suspend()`.

Filed as #214